subsys/pm/device_runtime.c

```diff
@@ -127,29 +124,48 @@ static int pm_device_request(const struct device *dev,
 						  PM_DEVICE_STATE_SUSPEND,
 						  NULL, NULL);
 		}
-		ret = 0;
 		goto out;
 	}
 
+	key = k_spin_lock(&dev->pm->lock);
+
+	if (!dev->pm->enable) {
+		ret = -ENOSYS;
+		goto out_unlock;
+	}
+
+	if (target_state == PM_DEVICE_STATE_ACTIVE) {
+		dev->pm->usage++;
+		if (dev->pm->usage < 0) {
+			goto out_unlock;
+		}
+	} else {
+		dev->pm->usage--;
+		if (dev->pm->usage > 1) {
+			goto out_unlock;
+		}
+	}
+
 	(void)k_work_schedule(&dev->pm->work, K_NO_WAIT);
 
 	/* Return in case of Async request */
 	if (pm_flags & PM_DEVICE_ASYNC) {
-		ret = 0;
-		goto out;
+		goto out_unlock;
 	}
 
-	k_mutex_init(&request_mutex);
-	k_mutex_lock(&request_mutex, K_FOREVER);
-	(void)k_condvar_wait(&dev->pm->condvar, &request_mutex, K_FOREVER);
-	k_mutex_unlock(&request_mutex);
+	k_spin_unlock(&dev->pm->lock, key);
+	pm_device_wait(dev, K_FOREVER);
```

Comment:
Is that correct understanding that we'll wait forever here?
Or is there a specific behavior with K_FOREVER (in which case, it would need to be clarified in function header)

Reply:
It will wait until the state transition finish. If there is no pending transition and the device is not in the middle of one it will return without wait.